### PR TITLE
Add branch name guidelines to CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by kfogel**
This is as discussed in the comments in issue #35, although that issue
is not about contributor documentation.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -9,7 +9,35 @@ you might also want to look at [DESIGN.md](DESIGN.md).
...
+[pull requests](https://help.github.com/articles/about-pull-requests/)
+(PRs) there.
+
+### Branching and Branch Names
+
+We do all development on lightweight branches, with each branch
+encapsulating one logical changeset (that is, one group of related
+commits).  Please try to keep changesets small and well-bounded: a
+branch should usually be short-lived, and should be turned into a PR,
+reviewed, and merged to `master` as soon as possible.
```
Comment (jasonaowen) [inline CONTRIBUTING.md]: Strongly agree! Maybe expand on "why"?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -9,7 +9,35 @@ you might also want to look at [DESIGN.md](DESIGN.md).
...
+    871-fix-provider-risk-score
+
+Everything after the issue number is just a reminder what the branch
+addresses.  Sometimes a branch may address only part of an issue, and
+that's fine: different branches can start with the same issue number,
+as long as the summary following the issue number indicates what part
+of the issue that particular branch addresses.
+
+If there is no issue number associated with a branch, then don't start
+the branch name with a number.
```
Comment (jasonaowen) [inline CONTRIBUTING.md]: I'm neutral on naming branches by starting with an issue number. The `Commit Messages` section, below, both describes a policy and the reasoning behind it, and I think this section would benefit from that kind of explanation as well. What is the problem this policy solves?

**Comment by jasonaowen** [comment]
One thing I would like, and have been trying to practice, is to include any relevant issue numbers and titles in the commit messages themselves. This helps when reviewing code long since merged (`git blame` -> find line -> `git show <commit>`), and including the issue title helps avoid a round-trip to the GitHub issue tracker. I think this somewhat addresses the same goal of branch-names-with-numbers, but might also be somewhat out of scope for this PR.

File: CONTRIBUTING.md (hunk 1 above)
Comment (cecilia-donnelly) [inline CONTRIBUTING.md]: Something like "Keeping branches short-lived reduces the likelihood of conflicts when it comes time to merge them back into master" maybe.

File: CONTRIBUTING.md (hunk 2 above)
Comment (cecilia-donnelly) [inline CONTRIBUTING.md]: I like it, but I'm also used to following this policy.  I find it helpful when I want to look at the branch list and immediately know which branch(es) has (have) work on a given issue.  I can do `git branch --all` and see that there are a couple branches open for work on issue `n`.  @kfogel, other thoughts?

Comment (jvasile) [inline CONTRIBUTING.md]: `git branch --all` at first looks like a maze of twisty branches, all the same.  Tiny bits of light at the start of each branch help keep the grue away.

File: CONTRIBUTING.md (hunk 1 above)
Comment (jvasile) [inline CONTRIBUTING.md]: I added @cecilia-donnelly's text to the PR.

**Comment by jvasile** [comment]
I'm merging this as I think it is helpful, even if it's not perfect.  

@jasonaowen is neutral on prefixing branches with issue numbers.  CMD, KFogel and I think it's a good idea.  I think the balance of opinion says make it a habit.

@jasonaowen added some additional thoughts about practices in commit messages.  I think those are helpful.  If you want to put that advice in CONTRIBUTING.md, go for it.

> One thing I would like, and have been trying to practice, is to include any relevant issue numbers and titles in the commit messages themselves. This helps when reviewing code long since merged (git blame -> find line -> git show <commit>), and including the issue title helps avoid a round-trip to the GitHub issue tracker. I think this somewhat addresses the same goal of branch-names-with-numbers, but might also be somewhat out of scope for this PR.

**Pull request merged**